keps/127: Support User Namespaces (#2101)
kubernetes/enhancements: keps/sig-node/127-usernamespaces-support/README.md

values:

- **Host**:
  The pods are placed in the host user namespace, this is the current Kubernetes
  behaviour. This mode is intended for pods that only work in the initial (host)
  user namespace. It is the default mode when `userNamespaceMode` field is not
  set.

- **Cluster**:
  All the pods in the cluster are placed in a _unique_ user namespace but they
  use the same ID mappings. This mode doesn't provide full pod-to-pod isolation
  as all the pods with `Cluster` mode have the same effective IDs on the host.
  It provides pod-to-host isolation as the IDs are different inside the
  container and in the host. This mode is intended for pods sharing volumes as
  they will run with the same effective IDs on the host, allowing them to read
  and write files in the shared storage.

- **Pod**:
  Each pod is placed in a different user namespace and has a different and
  non-overlapping ID mapping. This mode is intended for stateless pods, i.e.
  pods using only ephemeral volumes like `configMap,` `downwardAPI`, `secret`,
  `projected` and `emptyDir`. This mode not only provides host-to-pod isolation
  but also pod-to-pod isolation as each pod has a different range of effective

>   IDs in the host.

Author: Yes, that's right. I updated the text to make it explicit.

  smoother.
- **Duplicated Container Images Snapshots and Garbage Collection**:
  The `Pod` mode makes it more difficult to handle the [duplicated
  snapshots](#duplicated-snapshots-of-container-images) issue as it's possible
  that a pod uses a unique ID mapping each time it's scheduled. The different
  runtimes will have to use solutions like `metacopy` option of overlayfs or new
  kernel features to overcome it. It's also likely that the kubelet image garbage collection
  algorithm has to be changed as image snapshots shold be deleted as soon as the
  container finihes.
- **ID Mappings Allocation Algorithm**
  The `Pod` mode requires to have each pod in different and non-overlapping ID
  mapping. It requires to implement an algorithm that performs that allocation.
  There are some open questions about it:
    - What should be the length of the mapping assigned to each pod?
    - How to get the ID mapping range of a running pod when kubelet crashes?
    - Can the user specify the ID mappings for a pod?
- **High IDs in Container Images**:
  The IDs present on the image are not available as image metadata. The runtimes

>   would have to perform an image check, like analysing the `/etc/password` file,

Reviewer: ```suggestion
would have to perform an image check, such as analysing the `/etc/passwd` file,
```

> would have to perform an image check, such as analysing the `/etc/passwd` file,

Reviewer: D'oh sorry
```suggestion
  would have to perform an image check, such as analysing the `/etc/passwd` file,
```

`userNamespaceMode`.

#### Duplicated Snapshots of Container Images

The owners of the files of a container image have to been set accordingly to the
ID mappings used for that container. For example, if the user 0 in the container
is mapped to the host user 100000, then the `/root` directory has to be owned by
user ID 100000 in the host, so it appears to belong to root in the container.
The current implementation in container runtimes is to recursively perform a
`chown` operation over the image snapshot when it's pulled. This presents a risk
as it potentially increases the time and the storage needed to handle the
container images.

[containers/storage](https://github.com/containers/storage/) used by CRI-O
mounts an overlay file system with the
[`metacopy=on`](https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#metadata-only-copy-up)
flag set, it then chowns all of the lower files in the image to match the user
namespace to which the container will run. This operation is very quick compared
to standard chowning, since none of the files data has to be copied up. If a
second container runs on the same image with the same user namespace, then the

> chowned image is shared, eliminating the need to chown again.

Reviewer 2: Can we also mention uidmapping functionality in fuse-overlayfs? https://github.com/containers/fuse-overlayfs#usage
Supported by containerd as well as by CRI-O: https://github.com/containerd/containerd/blob/c623d1b36f09f8ef6536a057bd658b3aa8632828/snapshotter_opts_unix.go#L27-L35

Author: I added a reference to it below. Thanks for the pointers!

the full advantanges of user namespaces can be used in some cases (stateless
workloads).

There are some things that have to be studied with more detail for these
phase(s) but are not needed for phase 1, hence they are not discussed in detail:

- **Pod Default Mode**:
  It's not clear yet what should be the process to make this happen as this is a
  potentially non backwards compatible change. It's specially relevant for
  workloads not compatible with user namespaces. A [host defaulting
  mechanishm](#host-defaulting-mechanishm) can help to make this transition
  smoother.
- **Duplicated Container Images Snapshots and Garbage Collection**:
  The `Pod` mode makes it more difficult to handle the [duplicated
  snapshots](#duplicated-snapshots-of-container-images) issue as it's possible
  that a pod uses a unique ID mapping each time it's scheduled. The different
  runtimes will have to use solutions like `metacopy` option of overlayfs or new
  kernel features to overcome it. It's also likely that the kubelet image garbage collection
  algorithm has to be changed as image snapshots shold be deleted as soon as the

>   container finihes.

Reviewer 3: ```suggestion
  container finishes.
```

  The default size of the range for each container is `65536`. This can be tuned up
  based on the feedback we receive.
  There are some open questions about it:
    - Should be the size of the range configurable by the operator?
    - How to get the ID mapping range of a running pod when kubelet crashes?
    - Can the user specify the ID mappings for a pod?
- **High IDs in Container Images**:
  The IDs present on the image are not available as image metadata. The runtimes
  would have to perform an image check, such as analysing the `/etc/passwd` file,
  to discover what those IDs are. The kubelet and the CRI will require some
  changes to make this information available to the ID mappings allocator
  algorithm. It would have to be sure that the allocated mappings include those
  IDs and should have some logic to protect against special crafted images to
  perform a kind of DOS allocating too many IDs for a given container.
- **Security Considerations**:
  Once `Pod` is the default mode, it is needed to control who can use `Host` and
  `Cluster` modes. This can be done through Pod Security Policies if they are

>   available at thet time of implementing this phase.

Reviewer 3: ```suggestion
  available at the time of implementing this phase.
```

the container is the effective GID in the host when the fsGroup is defined. If
the `fsGroup` is part of the cluster ID mappings, a hole is "punched", otherwise
an extra one element ID mapping is added.

For instance, if the cluster GID mappings are

```
ContainerID HostID Size
0           1000   0
```
and there is a pod with `fsGroup: 500`, the GID mappings for that specific pod
are

```
ContainerID HostID Size
0           1000   500
501         1501   499
500         500    1
```

On the other hand, if a pod has `fsGroup: 3000`, the GID mappings for that pod are

```
ContainerID HostID Size
0           1000   1000
3000        3000   1
```

### Updating Ownership of Ephemeral Volumes

Ephemeral volumes use
[`AtomicWriter`](https://github.com/kinvolk/kubernetes/blob/master/pkg/volume/util/atomic_writer.go)
to create the files that are mounted to the containers. This component [has some
logic](https://github.com/kinvolk/kubernetes/blob/c94242a7b1d238cc27aea9b6d45ccb9963e814bb/pkg/volume/util/atomic_writer.go#L403)
to update the ownership of those files in some cases. It will be extended to
take the ID mappings into consideration when the pod runs in `Cluster` mode.

### Test Plan

> 

Reviewer 4: I know this is still in progess and provisional, but wanted you to know that once you move to implementable for a certain release (1.2x)  test plans will be required.

Author: Thanks!. Yes, I'd like to discuss first the technical details of the KEP and then I'll work on the test plan.

Reviewer 4: sounds perfect :+1:

  of the future unknown runtime and kernel vulnerabilities.

### Non-Goals

- Provide a way to run the kubelet process or container runtimes as an
  unprivileged process. Although initiatives like
  [usernetes](https://github.com/rootless-containers/usernetes) and this KEP
  both make use of user namespaces, it is a different implementation for a
  different purpose.

## Proposal

This proposal aims to support user namespaces in Kubernetes by extending the pod
specification with a new `userNamespaceMode` field. This field can have 3
values:

- **Host**:
  The pods are placed in the host user namespace, this is the current Kubernetes
  behaviour. This mode is intended for pods that only work in the initial (host)
  user namespace. It is the default mode when `userNamespaceMode` field is not
  set.

- **Cluster**:
  All the pods in the cluster are placed in a _unique_ user namespace but they
  use the same ID mappings. This mode doesn't provide full pod-to-pod isolation
  as all the pods with `Cluster` mode have the same effective IDs on the host.
  It provides pod-to-host isolation as the IDs are different inside the

>   container and in the host. This mode is intended for pods sharing volumes as

Reviewer 5: Flags on kubelet are much harder for users to manage, especially in hosted situations.  Is the intention to have people configure this flag based on their security posture or to aggressively move people to the more isolated model?

If the goal is to move towards more isolation, why expose it at all?  Just say that pods where `hostUserNamespace: false` will get a more isolated user namespace, as determined by the system.  Don't over-spec it.  For implementation, you can do something like serviceacct.uid % 2000000000 as a start.  Or we can simply add an allocation to serviceaccount and get it done right from the beginning.

>   container and in the host. This mode is intended for pods sharing volumes as

Author: Our intention now is to have this under people’s control. We’d like to enforce more isolation by default but unfortunately it’s not possible because the volume access problems when user namespaces are used. 

Providing user namespaces support in Kubernetes is a long effort and I don’t think it’s possible to have everything in a single iteration/KEP. This KEP is an initial effort trying to bring user namespaces support for some cases in Kubernetes. By having these fixed and non-overlapping allocation strategies we let the operator choose what the case they care the most, provide inter-pod isolation or support volumes.  

I feel we should not try to push for more sophisticated solutions than fixed allocation / cluster mode in this initial KEP.  Otherwise we risk to spend too much time trying to agree on a path for the more complicated use cases failing to even provide a solution for the simplest ones.

>   container and in the host. This mode is intended for pods sharing volumes as

Reviewer 5: I'm sensitive to not wanting to boil the ocean all at once, but I am not convinced we can't get closer in rev 1.

Specifically: flags to kubelet are a TERRIBLE interface - they simply are not accessible by a large fraction of users (in hosted services) and managed providers can't reasonably set them en masse. 

For compat, we can't change the default (yet).  I don't think anyone disagrees with that.  But I think we can cover 90% or more of use-cases with a per-NS or per-SA mapping as the obvious opt-in.   My main argument, though is that the act of opting-in can't be kubelet flags.

>   container and in the host. This mode is intended for pods sharing volumes as

Reviewer 6: Cool, so we are all at the same page regarding not changing the default :)

So, to clarify, moving the `mode` (cluster, pod) away from the pod spec seems like a good improvement to the KEP, right? Did I follow that correctly? In that case, the only pod spec change will be the useHostNamespace field (bool). I think this is way better, but want to be sure we are on the same page regarding that too.

What is yet not clear is how to decide a mapping for a pod. Let me give a little bit of context.

We proposed one strategy that is a “fixed” mapping, that is: each pod has its own user namespace but all pods share the same mapping. The advantage of this mode is that volumes can work (if container user 0 is mapped to user X on the host, the effective user ID in the files will be X and therefore can work with any other pod). However, Red Hat for example (cc @muranlp) wants mappings where non-overlapping ranges are given to pods. That is, no two pods are running in the same node with the same mappings. This isolates more pods between them, but volumes can’t reasonably work.

What I like about the fixed mapping is that it is super simple, works for MOST use cases (even sharing volumes with different namespaces) but gives you WAY more security that what we currently have without user namespaces. Seems like a net improvement, IMHO. Non-overlapping mappings gives you more security, but impossible to work with volumes (so, it is a different trade-off)

I’m not convinced about per-service-account mapping: if you want to share a volume, you also need to share a service account and if the pods are using the service account for something, now it needs to have the union of all the permissions? Seems too hacky and a trade-off in security (you use user ns, that is good thing security wise, but now your service accounts have more permissions than needed for some pods)

The per-namespace suggestion seems more reasonable, as it doesn't have the problems per-svca has. It of course has some limits (volume sharing per namespace only), but seems okay-ish. I’m sure these limits will be a problem in the future and will come back to bite us and I’m also not quite sure how the implementation of this would be (what a “cluster-unique” allocation for the mapping per ns looks like?).

In any case, if @mrunalp is fine with no support for non-overlapping mappings, I think we can discuss further what can be the only one mapping allocation to use. But I’d like to know what he thinks, as he really wanted non-overlapping mappings.

If we want to support different trade-offs (non-overlapping and some other way), does it make sense to move to a “strategy” configured in the kubelet (we then discuss what will be the strategies) as mauricio explained in this thread?

> For compat, we can't change the default (yet). I don't think anyone disagrees with that. But I think we can cover 90% or more of use-cases with a per-NS or per-SA mapping as the obvious opt-in. My main argument, though is that the act of opting-in can't be kubelet flags.

It probably makes sense to configure which ranges to use for any strategy (per-ns, fixed, non-overlapping or whatever), so one configuration in the Kubelet probably can be added. Or do you mean to have sensible default values for the mappings? Like use UID 10 millon or $SOMETHING_USUALLY_UNUSED on the host, as that is a good default for most systems and you can change it if you want? That SGTM too.

>   container and in the host. This mode is intended for pods sharing volumes as

Reviewer 5: > So, to clarify, moving the mode (cluster, pod) away from the pod spec seems like a good improvement to the KEP, right? Did I follow that correctly?                                                                                                                                         

I am not sure - I think we need to play out the rest to know the answer here.

> We proposed one strategy that is a “fixed” mapping, that is: each pod has its own user namespace but all pods share the same mapping.

This is utterly pragmatic, but it only addresses 1 requirement - protect the
node from pods.  Any attack that escapes one pod automatically compromises all
other pods on that node.  I feel like we can do better. 

> However, Red Hat for example (cc @mrunalp ) wants mappings where non-overlapping ranges are given to pods. That is, no two pods are running in the same node with the same mappings

IMO this is a step too far unless the FS-related issues are addressed.  As a
final goal, it sounds great.  But it suffers the same problems described
elsewhere - the inside space and the outside space are the same size, which
means you can't GUARANTEE that it works for any arbitrary 2^32 UID.

> What I like about the fixed mapping is that it is super simple, works for MOST use cases (even sharing volumes with different namespaces) but gives you WAY more security that what we currently have without user namespaces.

I can't agree with "WAY more".  A bit more, yes.

> I’m not convinced about per-service-account mapping: if you want to share a volume, you also need to share a service account and if the pods are using the service account for something, now it needs to have the union of all the permissions? Seems too hacky and a trade-off in security (you use user ns, that is good thing security wise, but now your service accounts have more permissions than needed for some pods)

I have a STRONG feeling that pods which are collaborating on a volume will tend
to have the same ServiceAccount.  I don't have hard data for that.  That said,
it feels like a relatively small increase in complexity over fixed mapping with
a low (IMO) risk of failure) and much better security.  The good news is that 
there's a clean "out" in the form of `hostUserNamespace` which we could advise
while we consider other options.

> I’m sure these limits will be a problem in the future and will come back to bite us_

The more "niche" the use-case, the less bad I feel about telling people to use
the host namespace :)

> and I’m also not quite sure how the implementation of this would be (what a “cluster-unique” allocation for the mapping per ns looks like?).

The *simple* answer is to take the UID of the SA (or the hash of the UID) %
(2^32 - 2) (because -1 is reserved) & ~(0xffff).  That gives us approximately
64K ID-spaces of 64K IDs each.  If we don't need that many IDs, We can shift it
to something like 4 million ID-spaces of 1024 Ids each.  Yes there's a chance
of collision, but low.  Yes, it still ignores the fact that any given pod could
use IDs outside the range we assign, regardless of what range we assign.

The *complete* solution involves new API resources and allocation and things
like that.  I am disinclined to devise the solution until we think we really
need it.

> I’d like to know what he thinks, as he really wanted non-overlapping mappings.

If we thought that a solution to the volumes problem was less than a year away,
we could take any one of these as a stepping-stone.  At that point, the less
surface area, the better.  Especially since this is opt-in and *any* solution
is better than what we have today.

> It probably makes sense to configure which ranges to use for any strategy (per-ns, fixed, non-overlapping or whatever), so one configuration in the Kubelet probably can be added. Or do you mean to have sensible default values for the ranges? Like use UID 10 millon or $SOMETHING_USUALLY_UNUSED on the host, as that is a good default for most systems and you can change it if you want? That SGTM too.

If we need a value that is consistent across machines, never changes, and is
not something a user would need to tweak - sure.  E.g. the overflow UID.  As
soon as we tell users "go tweak this to change how your users experience
kubernetes" or "set this to improve security", flags are just awful.

I think we should aspire to either a) totally automatic or b) exposed via the
API.

>   container and in the host. This mode is intended for pods sharing volumes as

Reviewer 7: > IMO this is a step too far unless the FS-related issues are addressed. As a
> final goal, it sounds great. But it suffers the same problems described
> elsewhere - the inside space and the outside space are the same size, which
> means you can't GUARANTEE that it works for any arbitrary 2^32 UID.

in CRI-O we punch holes for runAsUser/runAsGroup/fsGroup and map them back to their value on the host, so anything that relies on these values will still keep working.

We don't handle yet arbitrary IDs in the image files set to higher values than what is available in the inner user namespace, e.g. given an allocation of 65k IDs, a file owned by 200000:200000 will show up as owned by the overflow ID.

I think this issue could also be addressed in the runtime/kubelet.  The available IDs in the userns can be allocated in a non contiguous way to please the files present in the image.  What is preventing to address this issue right now in CRI-O is that the IDs must be allocated when the sandbox is created and the containers/images running in that pod are not known in this phase.

Another possibility would be to create a per-pod user namespace and then each container runs in their own sub user namespace, child of the pod user namespace; but that would probably require changes in the OCI runtime as well.

>   container and in the host. This mode is intended for pods sharing volumes as

Author: > in CRI-O we punch holes for runAsUser/runAsGroup/fsGroup and map them back to their value on the host, so anything that relies on these values will still keep working.

I'm curious, how does it work when the container run as root?

> I think this issue could also be addressed in the runtime/kubelet. The available IDs in the userns can be allocated in a non contiguous way to please the files present in the image. What is preventing to address this issue right now in CRI-O is that the IDs must be allocated when the sandbox is created and the containers/images running in that pod are not known in this phase.

I think there are two different scenarios. One is to guarantee that a mapping for the files on the container image exist, that's something that could be implemented. The other case are IDs on the fly that a container could try to use, that seems very complicated to cover IMO.

>   container and in the host. This mode is intended for pods sharing volumes as

Reviewer 7: > > in CRI-O we punch holes for runAsUser/runAsGroup/fsGroup and map them back to their value on the host, so anything that relies on these values will still keep working.
> 
> I'm curious, how does it work when the container run as root?

with the current annotations mechanism, root is treated as any other user so a container that specifies `runAsUser=0` will be mapped to root on the host, since that is the assumption from the outside.  It is possible though to use `runAsUser=` with a nonzero UID and specify to map it to root in the user namespace.

> > I think this issue could also be addressed in the runtime/kubelet. The available IDs in the userns can be allocated in a non contiguous way to please the files present in the image. What is preventing to address this issue right now in CRI-O is that the IDs must be allocated when the sandbox is created and the containers/images running in that pod are not known in this phase.
> 
> I think there are two different scenarios. One is to guarantee that a mapping for the files on the container image exist, that's something that could be implemented. The other case are IDs on the fly that a container could try to use, that seems very complicated to cover IMO.

I don't think the second case can be solved considering how user namespaces work now.  The assumption is that the container, even if running with CAP_SETUID/CAP_SETGID, will still be limited to use the IDs that are present in the image, either through files ownership or users defined in `/etc/passwd`.

If a container needs access to any of the 2^32 UIDs, then IMO it should not run in a user namespace.

>   container and in the host. This mode is intended for pods sharing volumes as

Reviewer 6: @thockin thanks! responses inline.

>>So, to clarify, moving the mode (cluster, pod) away from the pod spec seems like a good improvement to the KEP, right? Did I follow that correctly?

>I am not sure - I think we need to play out the rest to know the answer here.

Heh, okay, let’s wait. But if we don’t want to create a new API object to specify the mapping (as you mentioned and I agree not doing that now) and we want to reduce the surface we expose, I think this bool to use host userns or not should be there for sure, and probably not much more (maybe some other thing, but hopefully we can just enable userns and that is all you need to do).

But let’s see what we arrive at :)

> This is utterly pragmatic, but it only addresses 1 requirement - protect the node from pods. Any attack that escapes one pod automatically compromises all other pods on that node. I feel like we can do better.

Heh, yes, this is pragmatic indeed. There is probably some more isolation between pods as they run in different user namespaces (like can’t read the mem of a pod in a different userns even if /proc is exposed). But yeah, if a breakout happens, all pods in the node can be compromised.

The other upside non-mentioned of a fixed mapping is that it works with containers that need big/long mappings too, as we can allow almost all the 2^32 UID range inside the container. In contrast, in the case of per-ns/per-sa we need to _guess_ which mappings will be used and hope for the best or ask to use the host user namespace in those cases (or expose an API object in k8s with the ranges needed or have some metadata about the containers with the used UIDs, etc.).

Those images with high UIDs that can’t be easily guessed happen to be our use case too. Although I see that maybe it is not an issue for several users and the appeal of per-sa or per-ns.

>> However, Red Hat for example (cc @mrunalp ) wants mappings where non-overlapping ranges are given to pods. That is, no two pods are running in the same node with the same mappings

> IMO this is a step too far unless the FS-related issues are addressed. As a final goal, it sounds great. But it suffers the same problems described elsewhere - the inside space and the outside space are the same size, which means you can't GUARANTEE that it works for any arbitrary 2^32 UID.

Good point, fully agree. Until the kernel improves in some way (supports 64 bits kuids or something) and we can properly emulate userns of 2^32 UIDs, but not for now.

> I have a STRONG feeling that pods which are collaborating on a volume will tend to have the same ServiceAccount. I don't have hard data for that. That said, it feels like a relatively small increase in complexity over fixed mapping with a low (IMO) risk of failure) and much better security. The good news is that there's a clean "out" in the form of hostUserNamespace which we could advise while we consider other options.

Good point, yes, it seems a reasonable trade off regarding complexity and security. I’m unsure if they will share a SA, if you consider volumes that are typically written by many (like NFS), but I’m willing to find out :-D. I have a soft preference for per-ns (maybe being utterly pragmatic again?), but both very good ideas, thanks!

However, for our use case we have containers with high IDs that can’t be easily guessed. For that reason _some_ way to have a fixed mapping will be very useful for us, so we can specify one long mapping to be used by all.

If you think that any configuration knob, to support containers with a wide range of UIDs is a complete no-go, we would probably continue with the other options (as the patch we need to do for Kubernetes with a fixed mapping will be simpler than today). But if you think it is okay to allow that in some way in early revisions, I’d really like to know which ways you think might be okay for that :)

>> and I’m also not quite sure how the implementation of this would be (what a “cluster-unique” allocation for the mapping per ns looks like?).

> The simple answer is to take the UID of the SA (or the hash of the UID) % (2^32 - 2) (because -1 is reserved) & ~(0xffff). That gives us approximately 64K ID-spaces of 64K IDs each. If we don't need that many IDs, We can shift it to something like 4 million ID-spaces of 1024 Ids each. Yes there's a chance of collision, but low. Yes, it still ignores the fact that any given pod could use IDs outside the range we assign, regardless of what range we assign

Oh, good point! Yes, I like a simple solution like that: doesn’t guarantee, but is quite low chances of collision (at least for me, that I’m not an expert on that matter). Thanks!

> The complete solution involves new API resources and allocation and things like that. I am disinclined to devise the solution until we think we really need it.

I fully agree.

The only “downside” that I can find is not a big one: you can’t easily know the mapping in advance, so you can’t just add to your yaml “useHostUserNs: false” or you can’t easily migrate volumes to a new cluster (the UID will change), you will need to adjust (one time) your files permissions when switching to userns. Maybe this is even trivial as maybe most cases can use `fsGroupChangePolicy` when doing so and others can adjust as they need?

In any case, I don’t think that is a big problem and applies to any way of using userns with volumes without exposing the mapping. I agree we don’t want to expose the mapping, at least for now, so a downside we will have to live with (and not so bad, IMHO).

> If we thought that a solution to the volumes problem was less than a year away, we could take any one of these as a stepping-stone. At that point, the less surface area, the better. Especially since this is opt-in and any solution is better than what we have today.

I think it is really hard to know when the kernel will improve on that. But definitely, less surface makes sense and anything is better than what we currently have, fully agree on that :). There is ongoing work on the kernel front, hopefully[1].

> If we need a value that is consistent across machines, never changes, and is not something a user would need to tweak - sure. E.g. the overflow UID. As soon as we tell users "go tweak this to change how your users experience kubernetes" or "set this to improve security", flags are just awful.
> I think we should aspire to either a) totally automatic or b) exposed via the API.

Heh, I hear the point of hosted solutions, I wasn’t really aware of this (makes sense, but haven’t heard it before). I think what is coming out of these conversations is quite nice and achieves that. Thanks a lot for stepping in @thockin! :)

[1]: lwn article: https://lwn.net/Articles/837566/. V4 is out now, some weeks ago.

  they will run with the same effective IDs on the host, allowing them to read
  and write files in the shared storage.

- **Pod**:
  Each pod is placed in a different user namespace and has a different and

>   non-overlapping ID mapping across the node. This mode is intended for stateless pods, i.e.

Reviewer 8: We need to worry about `reclaimPolicy: Retain` in addition to multi-access.

How many stateful scenarios do we enable by using service account as the atom instead of pod (i.e. all pods running as a service account have the same mapping, the mapping is unique to the service account)? There are some other benefits to mapping based on service account too, like better uid information in the audit subsystem logs.

>   non-overlapping ID mapping across the node. This mode is intended for stateless pods, i.e.

Reviewer 8: Saw the discussion below about a per-(k8s-)namespace mode and IIUC similar benefits of per-service-account would apply to that.

>   non-overlapping ID mapping across the node. This mode is intended for stateless pods, i.e.

Reviewer 5: I'm not sure Retain matters so much - the volume is "done".  It's not saved for later use, it's pending some admin action to destroy it.

>   non-overlapping ID mapping across the node. This mode is intended for stateless pods, i.e.

Reviewer 5: Yes, this is why Namespace or ServiceAccount modes are so attractive :)

  pods using only ephemeral volumes like `configMap,` `downwardAPI`, `secret`,
  `projected` and `emptyDir`. This mode not only provides host-to-pod isolation
  but also pod-to-pod isolation as each pod has a different range of effective
  IDs in the host.

### User Stories

#### Story 1

As a cluster admin, I want run some pods with privileged capabilities because
the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
capability to grant me any extra privilege on the host.

#### Story 2

As a cluster admin, I want to allow some pods to run in the host user namespace
if they need a feature only available in such user namespace, such as loading a
kernel module with `CAP_SYS_MODULE`.

> ### Notes/Constraints/Caveats

Reviewer 9: Have you thought about devices' permissions (device mounts are set by device plugins in kubernetes)? AFAICS the problem space is similar to volumes but devices don't have the luxury of `fsGroups` we could use for GID mapping. The additional challenge with devices is that different devices have their own GIDs and we've also seen that certain groups (e.g., `video`) have different GID value depending on the node OS.

kubernetes/issues/92211 talks about the challenges with non-root users we've tried to solve. The primary goal is to allow containers with non-root user processes access devices. It looks at least `Host` mode is compatible with that. User namespaces was recently brought up and thus my questions here.

host change.

This proposal supports volumes without changing the user and group IDs and
leaves that problem for the user to manage. Developing Linux kernel features
such as shiftfs could allow different pods to see a volume with its own IDs but
it is not available yet. Among the possible future kernel solutions, we can
list:

- [shiftfs: uid/gid shifting filesystem](https://lwn.net/Articles/757650/)
- [A new API for mounting filesystems](https://lwn.net/Articles/753473/)
- [user_namespace: introduce fsid mappings](https://lwn.net/Articles/812221/)

In regard to this proposal, volumes can be divided in ephemeral and
non-ephemeral.

Ephemeral volumes are associated to a **single** pod and their lifecyle is
dependent on that pod. These are `configMap`, `secret`, `emptyDir`,
`downwardAPI`, etc. These kind of volumes can work with any of the three
different modes of `userNamespaceMode` as they are not shared by different pods
and hence all the processes accessing those volumes have the same effective user
and group IDs. Kubelet creates the files for those volumes and it can easily set
the file ownership too.

> Non-ephemeral volumes are more difficult to support since they can be persistent

Reviewer 10: For PVC, as it shares inside namespace, would it be good to have a new Namespaced mode, which every pod in a namespace shares the same uidmap.

Author: It's something that was proposed a while ago too https://github.com/kubernetes/community/pull/2595/files#diff-473657a51a9a6deb70241820b28201d0653f398f9f0736a47f0d23ca57864fa6R56. 

We don't see big benefits on this approach as the isolation is partially and pods inside the same namespace are still not isolated among them. It also complicates the design as there is a new mode where ranges have to be allocated to.

We'd prefer to avoid this mode and see if we are able to cover most user needs with the Cluster mode. If we realize it's really needed we could do a further iteration on it.

Author: I agree that "namespace" and "service-account" modes are interesting ideas. 

The main goal we're trying to accomplish with this KEP is to increase pod-to-host isolation. We know that increasing pod-to-pod isolation would be also great but by doing the former we'll protect the host at least. 

The most difficult part of supporting user namespaces are the volumes, the lack of an ID shift (or similar) mechanism in the Linux kernel makes it so difficult to share a volume among pods with different mappings. If such mechanism were implemented probably we're not having this conversation.

`ClusterMode` is an easy way to solve these volumes issues, yes, it's not a perfect solution but enables us to use user namespaces in some cases. With the new proposal we're doing in the other comment we're trying to keep this as simple as possible by enabling user namespaces in some basic cases and letting the possibility to cover more cases in future iterations. 

I think trying to cover too many cases from the beginning is very difficult, maybe we can start by the basic ones and in the meanwhile keep and eye on new kernel proposals like [id-mapped mounts](https://lists.linuxfoundation.org/pipermail/containers/2020-December/042820.html) that would definitely help with the more complicated ones.

Reviewer 11: I agree with starting with basic, broader protections given the difficulties and open questions of finer-grained approaches, and that it's a very positive step to improve Pod-Host isolation. This proposal seems like a good start on that.

@mauriciovasquezbernal One question I have (curiosity more than anything) is what risks are inherent to the Cluster mode? Part of the reason for protecting the host is to protect Pods from each other, how much can I do to other Pods if I'm in the same user ns? I assume (maybe naïvely) the same things as if I broke out of a container and was in host user ns today? There are ofc many other reasons to protect the host so I think Cluster mode is still valuable as a first step, but I am curious.

@thockin Can we imagine any scenarios where two Pods running as distinct service accounts would want to write to the same PVC? Maybe that's just a bad architectural choice and we should discourage it anyway...

Author: @mtaufen before answering, I want to clarify that the Cluster mode intends to use the same ID mapping but pods are in different user namespaces. Said that, if you are able to break out into the host user namespace you'll be able to read/write files of other pods, send signals, change scheduling parameters and other things described in the "User and group identifiers" section of https://man7.org/linux/man-pages/man7/credentials.7.html. It's a similar situation as today if you run pods as non-root and you break into the host: you'll have a process running as non-root on the host. On the other hand, if you run pods as root the situation would be very different, you'll have a root process in the host and will be able to compromise all pods too, user namespaces mitigate it.

Reviewer 11: Thanks, makes sense.

and shared by multiple pods. This proposal supports volumes with two different
strategies:

> - The `Cluster` mode makes it easier for pods to share files using volumes when

Reviewer 11: Is SA the right semantics though? Easy to imagine shared access to one resource by multiple principals, generally speaking. Might still be the convenient choice.

> - The `Cluster` mode makes it easier for pods to share files using volumes when

Reviewer 11: Problem space feels something like same-origin on the Web, K8s namespace might be the natural analogue...

Also worth thinking about cases where users have two clusters with the same set of namespaces and expect to move workloads between them - could we ensure that the mapping was consistently configured for both clusters, even across concurrent, potentially out-of-order additions/removals of namespaces?

> - The `Cluster` mode makes it easier for pods to share files using volumes when

Author: Perhaps we could derive the mapping from the namespace name, to guarantee it's the same in different clusters. However I don't know the security implications of allowing users to indirectly control the ranges by the namespace name. 

Cases like this are another reason why we're want to have the same mapping for all pods with volumes.

> - The `Cluster` mode makes it easier for pods to share files using volumes when

Reviewer 11: If we started with something like Cluster mode (which I think is a great first step), it theoretically blocks us from transparently migrating to Namespace mode in the future, since mappings would likely change, right? At that point do we basically have to wait for shiftfs to make things more secure without additional user intervention?

> - The `Cluster` mode makes it easier for pods to share files using volumes when

Reviewer 11: > I don't know the security implications of allowing users to indirectly control the ranges by the namespace name.

It _might_ not be a huge problem, given namespaces are probably created by cluster admins, controlling ranges might fit their privilege level anyway.

> - The `Cluster` mode makes it easier for pods to share files using volumes when

Author: > If we started with something like Cluster mode (which I think is a great first step), it theoretically blocks us from transparently migrating to Namespace mode in the future, since mappings would likely change, right? 

Yes, users will have to update the volumes ownership. Btw, one of our goals with this proposal is to have something that works good for now and that allows to easily use a shiftfs-like solution when it'll be available. In one of our comments (https://github.com/kubernetes/enhancements/pull/2101#issuecomment-754778887) we're proposing to have a "Cluster" mode but implicit, to avoid changes to the pod spec. 

> At that point do we basically have to wait for shiftfs to make things more secure without additional user intervention?

Yes, I think shiftfs (or a similar solution) is the only way to avoid user intervention at all.

> - The `Cluster` mode makes it easier for pods to share files using volumes when

Reviewer 11: Ok, thanks for the clarification.

combinations based on their specific needs. For instance, if a pod access a
shared volume containing files and folders with permissions for `others`, it can
run in `Pod` mode. On the other hand, a process inside a pod will not be able to
access files with mode `0700` and owned by a user different than the effective
user of that process in a volume that doesn't support the semantics of `fsGroup`
(doesn't support
[`SetVolumeOwnership`](https://github.com/kubernetes/kubernetes/blob/00da04ba23d755d02a78d5021996489ace96aa4d/pkg/volume/volume_linux.go#L42)
that updates permissions and ownership of the files to be accesible by the
`fsGroup` group ID). Such pods should be run in `Host` mode.

#### Container Runtime Support

- **Docker**:
  Docker only supports a [single ID
  mappings](https://docs.docker.com/engine/security/userns-remap/) shared by all
  containers running in the host. There is not support for [multiple ID
  mappings](https://github.com/moby/moby/issues/28593) yet. Updating dockershim
  is out of scope of this KEP as [it's going to be
  deprecated]((https://github.com/kubernetes/enhancements/pull/1985/)) and it

>   offers a very limited support for user namespaces.

Reviewer 12: How will this impact people who have not moved off of Docker? I get that `Cluster` and `Pod` mode won't work with Docker, but `Host` mode will just work like this are today?

Reviewer 6: Yes, host mode will work just fine and with no changes needed

  The list of such capabilities is likely to change from one Linux version to
  another. For example, Linux now has [time
  namespaces](https://man7.org/linux/man-pages/man7/time_namespaces.7.html) and
  there are ways to make `CAP_SYS_TIME` work inside a user namespace. There are
  also discussions to make `CAP_MKNOD` work in user namespaces.

- **Sharing Host Namespaces**:
  There are some limitations in the Linux kernel and in the runtimes that
  prevent sharing other host namespaces when the host user namespace is not
  shared.
  - Mounting `mqueue` (`/dev/mqueue`) is not allowed from a process in a user
    namespace that does not own the IPC namespace. Pods with `hostIPC=true` and
    `userNamespaceMode=Pod|Cluster` can fail.
  - Mounting `procfs` (`/proc`) is not allowed from a process in a user
    namespace that does not own the PID namespace. Pods with `hostPID=true` and
    `userNamespaceMode=Pod|Cluster` can fail.
  - Mounting `sysfs` (`/sys`) is not allowed from a process in a user namespace
    that does not own the network namespace. Impact: pods with
    `hostNetwork=true` and `userNamespaceMode=Pod|Cluster` can fail.

>   If users specify `userNamespaceMode=Pod|Cluster` and one of these

Reviewer 5: Should we statically enforce this at the API layer?  Do we expect it to change over time?  Or should we just document all of these as "consult your OS docs to learn more about how these interact" ?

  `host{IPC,PID,Network}=true` options, runc will currently fail to start the
  container. The kubelet does **not** try to prevent that combination of
  options, in case runc or the kernel make it possible in the future to use that
  combination.

In order to avoid breaking existing workloads `Host` is the default value of
`userNamespaceMode`.

#### Duplicated Snapshots of Container Images

The owners of the files of a container image have to been set accordingly to the
ID mappings used for that container. For example, if the user 0 in the container
is mapped to the host user 100000, then the `/root` directory has to be owned by
user ID 100000 in the host, so it appears to belong to root in the container.
The current implementation in container runtimes is to recursively perform a
`chown` operation over the image snapshot when it's pulled. This presents a risk

> as it potentially increases the time and the storage needed to handle the

Reviewer 5: and argues for a more stable mapping

namespace to which the container will run. This operation is very quick compared
to standard chowning, since none of the files data has to be copied up. If a
second container runs on the same image with the same user namespace, then the
chowned image is shared, eliminating the need to chown again.

[fuse-overlayfs](https://github.com/containers/fuse-overlayfs) is a fuse-based
overlayfs implementation that supports ID shifting. It doesn't require a
recursive chown operation (avoiding the file duplication) but given that it's
implemented in user space has lower I/O performance.

More sophisticated approaches to this problem are being
[discussed](https://lists.linuxfoundation.org/pipermail/containers/2020-September/042230.html)
in the kernel community. This is something that container runtimes should use
once it's available and it does not impact the kubelet nor the CRI gRPC spec.

#### Container Images with High IDs

There are container images designed to run with high user and group IDs. It's
possible that the IDs range assigned to the pod is not big enough to accommodate
these IDs, in this case they will be mapped to the `nobody` user in the host.

> It's not a big problem in the `Cluster` case, the users have to be sure that

Reviewer 5: I'm a little confused about ranges - what are they used for and why are they needed?

Reviewer 5: why do we need ranges for pods?  Why not a single "outside" UID?

Author: It's possible that a pod uses multiple IDs inside. It's not possible to map all of those to the same "outside" ID but we need a range. If we only map one ID, then the other IDs will be seen as `nobody - 65534 ` from the host point of view witch creates issues to access files that don't have "others" permissions.

Reviewer 6: @thockin It is implicit on what @mauriciovasquezbernal is saying but the Linux kernel forces you to have a range (you need a 1-1 mapping, a bijection).

I'd like the same you mention (map UIDs/GIDs in the container to just ID X on the host), but it is not possible to do that in Linux.

Reviewer 5: That's simply impossible to do with any guarantee.  Each pod *might* use any UID from the entire space.  More concretely, they could use UID 0 and UID 2^32-1 - requiring the whole range to map.  The only way to guarantee non-collision on the outside is to limit which UIDs can be used on the inside to a "reasonable" range.

This seems like a pretty fundamental thing we need to design for, no?

Sorry if I am rehashing old discussion - I don't see this really covered here.  Are we assuming some "standard" range size for every pod?  Do pods get to express "I need more UIDs"?

One of the very fundamental principles of kubernetes is that users should not be subject to collisions that they did not cause.  For example, sharing host ports.

Hypothetically: Assume we had to pre-allocate "outside" UIDs and the only UIDs you could use were those - would it be better?  What would the unit of allocation-assignment?  I think Namespace or ServiceAccount seems right.  So if we carry that to the hypothetical end, what if we end up at a model like this:

* ServiceAccount gets new "users[]" and "groups[]" fields.  These describe the "inside" and are optional.
* An IDAllocation resource.  This is created in response to the above fields and assigns "outside" IDs to them.
* When a pod runs as that SA it gets that UID mapping.
* When the cluster runs out of IDs, no new IDAllocations can be made.
* Allocation must be pluggable.

That's really the only way to ensure no outside collision.  Is that a worthwhile goal?  Or is the better goal to buy time until node-local ephemeral mappings are viable wrt filesystems?  We still have the questions above - standard range size or configurable, etc.

Author: > Sorry if I am rehashing old discussion - I don't see this really covered here. Are we assuming some "standard" range size for every pod? Do pods get to express "I need more UIDs"?

Yes, 65k (UID 0 -> UID 2^16-1) seems to be that [reasonable default range size](https://github.com/kubernetes/enhancements/pull/2101#issuecomment-712960297) (https://github.com/kubernetes/enhancements/pull/2101/files#diff-2d23f433c4d58d94c34afbadc193009e2f7b4b8eab1c6549f7bd96d81abf142bR399-R407). It should be enough for a pod and also allows to schedule up to 65k pods in a node (far away of what k8s handles these days). 

We haven't considered a case where a pod can ask for more UIDs. We'd prefer to see a use case that needs it before proposing something similar.

> One of the very fundamental principles of kubernetes is that users should not be subject to collisions that they did not cause. For example, sharing host ports.

I'm not sure what it means for k8s, but with user namespaces the quantity of IDs available for each pod will be reduced because it's a node-wide shared resource. We'll need to limit the inside range of UIDs a pod can use to something like 0 -> 65k, at least for rev 1. If a pod needs to use IDs outside that range it could share the host user namespace anyway. 

> Hypothetically: Assume we had to pre-allocate "outside" UIDs and the only UIDs you could use were those - would it be better? What would the unit of allocation-assignment? I think Namespace or ServiceAccount seems right. So if we carry that to the hypothetical end, what if we end up at a model like this:

I think you described the allocation algorithm in a high level detail. In general this algorithm should have a pool of host (outside) IDs it can use (maybe configurable by the operator?), each time an "allocation unit resource" (pod / serviceaccount / namespace depending on the outcome of our discussion) is created, this algorithm assigns a portion of that range to that resource.  

> ServiceAccount gets new "users[]" and "groups[]" fields. These describe the "inside" and are optional.

Do you mean to let the user control the "inside" part of the IDs mapping?

> That's really the only way to ensure no outside collision. Is that a worthwhile goal? Or is the better goal to buy time until node-local ephemeral mappings are viable wrt filesystems? We still have the questions above - standard range size or configurable, etc.

That's a good question. What do you mean with "node-local ephemeral mapping"? 
I think we should guarantee that each allocation unit gets a different and non-overlapping range. If we allocate per pod, each kubelet can make this decision, if we allocate per-namespace or per-serviceaccount we'll have to make this decision to an upper layer.

Reviewer 5: >> Sorry if I am rehashing old discussion - I don't see this really covered here. Are we assuming some "standard" range size for every pod? Do pods get to express "I need more UIDs"?
> 
> Yes, 65k (UID 0 -> UID 2^16-1) seems to be that reasonable default range size

Is that really necessary, given that most pods use exactly 1 UID?

> allows to schedule up to 65k pods in a node

Only if you assume every range is localized to a node.  If this was per-SA or
per-NS then it limits you to 64K SAs or NSes, which is actually something
people hit.

What if we gave everyone a smaller set of IDs by default, e.g.
0-1023 + 65520-65535.  Then, anyone who really needed more that that would
have to ask for it, and those special cases could get 64K allocations.

Just brainstorming.

> We haven't considered a case where a pod can ask for more UIDs. We'd prefer to see a use case that needs it before proposing something similar.

I know some apps do run as ID > 64K.  It's not that they need *MORE* IDs,
they'd need *different* ones.

> We'll need to limit the inside range of UIDs a pod can use to something like 0 -> 65k

That's a breaking change for some users.  They need a clean opt-out at least.
That's what the host space is for.  I agree.

>> ServiceAccount gets new "users[]" and "groups[]" fields. These describe the "inside" and are optional.
>
> Do you mean to let the user control the "inside" part of the IDs mapping?

If we end up going as far as an allocator, sure why not?

>> That's really the only way to ensure no outside collision. Is that a worthwhile goal? Or is the better goal to buy time until node-local ephemeral mappings are viable wrt filesystems? We still have the questions above - standard range size or configurable, etc.
> 
> That's a good question. What do you mean with "node-local ephemeral mapping"?

I meant shiftfs or whatever will make FS permissions compatible with user
namespaces.  It sounds like linux is as brain-damaged now as it was when I last
looked at userns - years ago. :(

> I think we should guarantee that each allocation unit gets a different and non-overlapping range. If we allocate per pod, each kubelet can make this decision, if we allocate per-namespace or per-serviceaccount we'll have to make this decision to an upper layer.

I agree with your conclusion, I am less convinced on your predicate.  I am not
sure we *NEED* to *guarantee* that each allocation unit gets a different and
non-overlapping range.  If we simply ensure a low probability of collision, we
can buy enough time for a "real" solution.

I don't know which solution will be "real" enough or how we would migrate
users, but I am sure we can figure that out...

Author: > For instance, if a pod gets the `0 100000 65536` mappings and a container inside the pod has `USER: 200000`, it'll be running as `nobody` on the host because the range of length 65536 is not long enough to map this ID to the host.

Ops, I was wrong on this statement. If the user the container should run as is not part of the mapping it'll fail to start. Also calls to `setuid/setgid` within the container trying to set an ID not present on the mapping will fail as well. 

In other words, any container trying to run with an UID not present on the mapping will have a hard failure.

Author: > Is that really necessary, given that most pods use exactly 1 UID?

Well, with 64K we’ll cover even more cases. If we go for a per namespace / svc account approach we’ll need to reduce that range as you mentioned above. Finding the right size is important, the bigger the range the less allocations we can do and the higher the probability of a collision, on the other hand, the smaller the range the more applications won’t work. 

> Then, anyone who really needed more that that would
> have to ask for it, and those special cases could get 64K allocations.

I think it’s related to my question about letting the user configure the inside part of the mapping. Probably it’s something we don’t want to do for rev1, asking the user to share the host namespace could be enough for now. For future revisions we could let the user choose the inside part of the mapping, so the IDs they’re using are mapped. 

> I know some apps do run as ID > 64K. It's not that they need _MORE_ IDs,
> they'd need _different_ ones.

Oh yeah, exactly. 

> I meant shiftfs or whatever will make FS permissions compatible with user
> namespaces. It sounds like linux is as brain-damaged now as it was when I last
> looked at userns - years ago. :(

Yes, we don’t have more solutions now. shiftfs will probably never be merged. There is some ongoing work by Christian Brauner[1] but unsure what will happen and, assuming the best case that it is merged and when that kernel will be available in distros

[1]: lwn article: https://lwn.net/Articles/837566/. V4 is out now, some days ago (lwn links to v1 or v2)
 
> I agree with your conclusion, I am less convinced on your predicate. I am not
> sure we _NEED_ to _guarantee_ that each allocation unit gets a different and
> non-overlapping range. If we simply ensure a low probability of collision, we
> can buy enough time for a "real" solution.

I’m not sure either. Having non-overlapping ranges now that we can’t guarantee that each pod runs in a different mapping could be an overkill. I agree we can buy time by having a simpler approach, however I’d like to know the opinion of the Red Hat folks as we know they want to have non-overlapping mappings @mrunalp @giuseppe @rhatdan

  // "Cluster": The pod uses a cluster-wide configured ID mappings.
  // +k8s:conversion-gen=false
  // +optional
  UserNamespaceMode PodUserNamespaceMode `json:"userNamespaceMode,omitempty" protobuf:"bytes,36,opt  name=userNamespaceMode"`
...
```

#### Option 2: PodSecurityContext

Add it to `v1.PodSpec.PodSecurityContext` as this field controls a security
aspect of the pod.

```
const (
	UserNamespaceModeHost    PodUserNamespaceMode = "Host"
	UserNamespaceModeCluster PodUserNamespaceMode = "Cluster"
)

type PodSecurityContext struct {
...
  // UserNamespaceMode controls how user namespaces are used for this Pod.
  // Three modes are supported:
  // "Host": The pod shares the host user namespace. (default value).
  // "Cluster": The pod uses a cluster-wide configured ID mappings.
  // +optional
  UserNamespaceMode PodUserNamespaceMode `json:"userNamespaceMode,omitempty" protobuf:"bytes,11,opt  name=userNamespaceMode"`
...
```
<<[/UNRESOLVED]>>

### Configuring the Cluster ID Mappings

This proposal considers two different ways to configure the ID mappings used for
the `Cluster` mode. This is for discussion with the community and only one will
be considered.

> <<[UNRESOLVED where to configure the cluster wide ID mappings ]>>

Reviewer 5: We'll need to think through how this can change over time.  I know you want to say "don't change it" but experience says it will eventually need to be changed.,

overlapping mappings, to prevent kubelet sending wrong ID mappings to the
runtime.

The following is an example a configuration file:

```
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
clusterIDMappings:
  uidMappings:
  - containerID: 0
    hostID: 500000
    size: 65536
  - containerID: 100000
    hostID: 600000
    size: 65536
  gidMappings:
  - containerID: 0
    hostID: 500000
    size: 65536

```

**Pros**:
 - It's directly accessible by the kubelet. It's not needed to expose through an
   API and there are not concerns if the api server is down.
 - No modifications are needed to other components.

**Cons**:
 - It's more difficult for the user as they have to guarantee the same ID
   mapping is configured in all the nodes.

#### Option 2: Configure as a Cluster Parameter in kube-apiserver

This option considers setting this parameter on the kube-apiserver.

**Pros**:
 - Easier for the user as the parameter is configured once in a single place.

 **Cons**:
 - It's difficult to expose this parameter to the kubelet.
 - The parameter could not be available for the kubelet if the kube-apiserver is
   down.

> 

Reviewer 5: Option 3 - stick it in Node
Option 4 - make a new resource which is 1:1 with Nodes for this sort of info

<<[/UNRESOLVED]>>

> ### 1-to-1 Mapping for fsGroup

Reviewer 5: Is this still an issue in namespace or svcacct mode?

> The `fsGroup` is asssumed to be the correct group owner of the files present on

Reviewer 7: typo in `asssumed`

Author: Thanks!
